lib/redis-session-store.rb

```diff
@@ -100,107 +68,120 @@ def verify_handlers!
   end
 
   def prefixed(sid)
+    return nil unless sid && sid.private_id
+    "#{default_options[:key_prefix]}#{sid.private_id}"
+  end
+
+  def prefixed_fallback(sid)
     "#{default_options[:key_prefix]}#{sid}"
   end
 
-  def session_default_values
-    [generate_sid, USE_INDIFFERENT_ACCESS ? {}.with_indifferent_access : {}]
+  def create_sid(req)
+    req.env['redis_session_store.new_session'] = true
+    generate_sid
   end
 
-  def get_session(env, sid)
-    sid && (session = load_session_from_redis(sid)) ? [sid, session] : session_default_values
-  rescue Errno::ECONNREFUSED, Redis::CannotConnectError => e
-    on_redis_down.call(e, env, sid) if on_redis_down
-    session_default_values
+  def find_session(req, sid)
+    with_redis_connection(default_rescue_value: [nil, {}]) do |redis_connection|
+      existing_session = load_session_from_redis(redis_connection, req, sid)
+      return [sid, existing_session] unless existing_session.nil?
+
+      [create_sid(req), {}]
+    end
   end
-  alias find_session get_session
 
-  def load_session_from_redis(sid)
-    data = with_redis do |redis|
-      redis.get(prefixed(sid))
+  def load_session_from_redis(redis_connection, req, sid)
+    return nil unless sid
+    if read_fallback
+      data = redis_connection.get(prefixed(sid)) || redis_connection.get(prefixed_fallback(sid))
+    else
+      data = redis_connection.get(prefixed(sid))
     end
 
     begin
       data ? decode(data) : nil
     rescue StandardError => e
-      destroy_session_from_sid(sid, drop: true)
+      delete_session_from_redis(redis_connection, sid, req, { drop: true })
       on_session_load_error.call(e, sid) if on_session_load_error
       nil
     end
   end
 
   def decode(data)
-    session = serializer.load(data)
-    USE_INDIFFERENT_ACCESS ? session.with_indifferent_access : session
+    serializer.load(data)
   end
 
-  def set_session(env, sid, session_data, options = nil)
-    expiry = get_expiry(env, options)
-    with_redis do |redis|
-      if expiry
-        redis.setex(prefixed(sid), expiry, encode(session_data))
-      else
-        redis.set(prefixed(sid), encode(session_data))
-      end
+  def write_session(req, sid, session_data, options = nil)
+    return false unless sid
+
+    if write_fallback
+      key = prefixed_fallback(sid)
+    else
+      key = prefixed(sid)
     end
-    sid
-  rescue Errno::ECONNREFUSED, Redis::CannotConnectError => e
-    on_redis_down.call(e, env, sid) if on_redis_down
-    false
-  end
-  alias write_session set_session
+    return false unless key
 
-  def get_expiry(env, options)
-    session_storage_options = options || env.fetch(ENV_SESSION_OPTIONS_KEY, {})
-    session_storage_options[:ttl] || session_storage_options[:expire_after]
+    set_options = {
+      ex: options[:expire_after] || default_redis_ttl
+    }
+
+    if req.env['redis_session_store.new_session'] == true
+      set_options[:nx] = true
+    end
```

Comment:
are there every truthy-but-not-true values for `'redis_session_store.new_session'`?

I guess I'm asking if we could simplify the check:
```suggestion
    if req.env['redis_session_store.new_session']
      set_options[:nx] = true
    end
```

Reply:
yep, we can